### DISCUSSION TITLE
Alternative Electron setup method

### DISCUSSION THREAD

**Discussion by goenning** [aptabase/aptabase-electron#1] [General]
There are many ways of setting up an Electron app. While our [recommended](https://github.com/aptabase/aptabase-electron) setup should work on any modern setup, it may fail to work on older versions of Electron.

If the recommended setup doesn't work for you, try this:

## Install

Install the SDK using your preferred JavaScript package manager

```bash
npm add @aptabase/electron
```

## Main Setup

On your `main` script, initialize the SDK before the app is ready:

```js
const { initialize } = require("@aptabase/electron");

initialize("<YOUR_APP_KEY>"); // 👈 this is where you enter your App Key

app.whenReady().then(() => {
  // ... the rest of your app initialization code
});
```

## Preload Setup

On your `preload` script, expose trackEvent on your window object.

You can use `contextBridge` if that's supported on your electron version.

```js
const { contextBridge, ipcRenderer } = require('electron')

contextBridge.exposeInMainWorld(
  'aptabase', 
  {
    trackEvent: (...args) => ipcRenderer.send("aptabase:trackEvent", ...args)
  }
)
```

Alternatively, if it doesn't, you can expose it directly on the window object

```js
const { ipcRenderer } = require("electron");

window.aptabase = {
    trackEvent: (...args) => ipcRenderer.send("aptabase:trackEvent", ...args)
}
```

## Usage

You can now trackEvents from your `main` process by using the `@aptabase/electron`:

```js
const { trackEvent } = require("@aptabase/electron");
...
trackEvent("something_happened")
```

In your client code, use the window object:

```js
window.aptabase.trackEvent("button_clicked", { name: "Submit" });
```

**Comment by vbhakta8** [comment]
Our installation worked with the recommend setup but stopped working randomly and we had to switch to exposing it directly to the window object. Any reason why we would have to do that?
